### ISSUE TITLE
Codes for coordinate generation

### ISSUE THREAD

**Issue by SamitHuang** [baidu-research/NCRF#29]
Could you release the codes for coordinates generation? Now the patch size is fixed to be 768x768. Once it changes, we need to generate new coordinates. Also need to extract and save a new set of extracted images, which is spatially expensive.

**Comment by yil8** [comment]
@SamitHuang The code for generating the coordinates are not prepared yet. It's a little bit messy. You can refer to this #14 . But I think you don't need to generate new coordinates when you change the patch size. Can you just use the current coordinates to generate a different patch size, e.g. 512x512?

**Comment by SamitHuang** [comment]
But if the patch size is larger than 768, re-checking tumor region is required.

**Comment by usamabaig1** [comment]
Google labeled the patch as a tumor even if one of the pixels is of tumor region. so i think we can also do same.

**Comment by yil8** [comment]
@SamitHuang @usamabaig1 So there are two concepts here, first is the grid size, e.g. 768, the other is the patch size within the grid, e.g. 256. Each patch within the grid is modeled through CNN and CRF. We obtained the tumor/normal label of each patch within the grid by checking the center coordinate of each patch with respect to the ground truth annotation mask. Therefore, for a grid size of 768 of 3x3 256 patches, the label will also be 3x3 of 1/0. If you just want to randomly sample tumor or normal coordinates, the code is pretty much already there by using the tissue mask code as noted in #14 . But the performance won't be great without hard negative mining. However the code for hard negative mining is quite a bit painful.

**Comment by SamitHuang** [comment]
Thanks. I understand the concepts. can you describe the hard negative mining procedure? Do you apply a previously trained CNN on each WSI for training exhaustively to find out all false negative tissue patches, then add them to the training set?

**Comment by yil8** [comment]
@SamitHuang Yes, I first trained a CNN on purely randomly sampled normal/tumor patches. Then applied this CNN on each tumor WSI in training set to find out all false negative tissue patches, then add them to the training set.
